trace/trace.go

```diff
@@ -56,6 +56,14 @@ func (s *Span) IsRecordingEvents() bool {
 	return s.data != nil
 }
 
+func (s *Span) ensureNotRecordingEvents() {
```

Comment:
I wonder if this could be merged with a version of makeSpanData to "steal" the SpanData instead of making a copy. AFAICT makeSpanData is only called in End in order to prevent concurrent modification through the Span.data pointer, but if we nil that pointer we don't need to worry about it any more and maybe there is no longer a need to make a deep copy?